tiledb/sm/array_schema/test/unit_attribute.cc

```diff
@@ -0,0 +1,207 @@
+  CHECK(a.cell_val_num() == constants::var_num);
+}
+
+// long form constructor spot-check
+TEST_CASE(
+    "ANY Attribute long form constructor invalid cell val num",
+    "[array_schema][attribute]") {
+  FilterPipeline empty_pipeline;
+  ByteVecValue empty_fill;
+  CHECK_THROWS(Attribute(
+      "a",
+      Datatype::ANY,
+      true,
+      1,
+      empty_pipeline,
+      empty_fill,
+      0,
+      DataOrder::INCREASING_DATA,
+      std::optional<std::string>()));
+}
+
+TEST_CASE(
+    "STRING_ASCII Attribute default cell val num",
+    "[array_schema][attribute]") {
+  Attribute a("a", Datatype::STRING_ASCII, false);
+  CHECK(a.cell_val_num() == 1);
+}
+
+// set at construction, non-var is not fine when ordered
+TEST_CASE(
+    "STRING_ASCII Attribute non-default cell val num ordered",
+    "[array_schema][attribute]") {
+  CHECK_THROWS(
+      Attribute("a", Datatype::STRING_ASCII, 1, DataOrder::INCREASING_DATA));
+  CHECK_THROWS(
+      Attribute("a", Datatype::STRING_ASCII, 100, DataOrder::INCREASING_DATA));
+}
+
+// set later, 1 is not fine when un-ordered
+TEST_CASE(
+    "STRING_ASCII Attribute set_cell_val_num ordered",
+    "[array_schema][attribute]") {
+  Attribute a(
+      "a",
+      Datatype::STRING_ASCII,
+      constants::var_num,
+      DataOrder::INCREASING_DATA);
+  CHECK_THROWS(a.set_cell_val_num(1));
+  CHECK_THROWS(a.set_cell_val_num(100));
+  CHECK(a.cell_val_num() == constants::var_num);
+}
+
+// long form constructor spot-check
+TEST_CASE(
+    "STRING_ASCII Attribute long-form constructor invalid cell val num",
+    "[array_schema][attribute]") {
+  FilterPipeline empty_pipeline;
+  ByteVecValue empty_fill;
+  CHECK_THROWS(Attribute(
+      "a",
+      Datatype::STRING_ASCII,
+      true,
+      1,
+      empty_pipeline,
+      empty_fill,
+      0,
+      DataOrder::INCREASING_DATA,
+      std::optional<std::string>()));
+}
+
+/**
+ * Other datatypes can be any size fixed or var unless ordered
+ */
+TEST_CASE("INT32 Attribute default cell val num", "[array_schema][attribute]") {
+  CHECK(Attribute("a", Datatype::INT32, false).cell_val_num() == 1);
+}
+
+TEST_CASE(
+    "INT32 Attribute non-default cell val num unordered",
+    "[array_schema][attribute]") {
+  CHECK(
+      Attribute("a", Datatype::INT32, 1, DataOrder::UNORDERED_DATA)
+          .cell_val_num() == 1);
+  CHECK(
+      Attribute("a", Datatype::INT32, 100, DataOrder::UNORDERED_DATA)
+          .cell_val_num() == 100);
+  CHECK(
+      Attribute(
+          "a", Datatype::INT32, constants::var_num, DataOrder::UNORDERED_DATA)
+          .cell_val_num() == constants::var_num);
+}
+
+TEST_CASE(
+    "INT32 Attribute set cell val num unordered", "[array_schema][attribute]") {
+  Attribute a("a", Datatype::INT32, false);
+  a.set_cell_val_num(1);
+  CHECK(a.cell_val_num() == 1);
+  a.set_cell_val_num(10);
+  CHECK(a.cell_val_num() == 10);
+  a.set_cell_val_num(constants::var_num);
+  CHECK(a.cell_val_num() == constants::var_num);
```

Comment:
Is order independence of these operations being verified? If so, it needs documentation.
Alternatively, are postconditions of `set_cell_val_num` being verified? If so, it needs three `SECTION`s.